feat: add search options for string encoding and endianness

PerikiyoXD opened the pull request:
Added search options for string encoding (UTF-8, UTF-16, UTF-32) and endianness (Little, Big) in the hex editor. This enhancement allows users to customize the search process based on different string encodings and byte orders.

Affected files:
- `plugins/builtin/romfs/lang/de_DE.json`
- `plugins/builtin/romfs/lang/en_US.json`
- `plugins/builtin/romfs/lang/es_ES.json`
- `plugins/builtin/romfs/lang/it_IT.json`
- `plugins/builtin/romfs/lang/ja_JP.json`
- `plugins/builtin/romfs/lang/ko_KR.json`
- `plugins/builtin/romfs/lang/pt_BR.json`
- `plugins/builtin/romfs/lang/zh_CN.json`
- `plugins/builtin/romfs/lang/zh_TW.json`
- `plugins/builtin/source/content/views/view_hex_editor.cpp`

Resolves: #1325

PerikiyoXD commented:
Should be ready to ship!

WerWolv commented on plugins/builtin/include/content/popups/hex_editor/popup_hex_editor_find.hpp:
```diff
@@ -0,0 +1,52 @@
+        UTF32
+    };
+
+    enum class StringModeEndianness : u8 {
```
I generally use `std::endian` for this

WerWolv commented on plugins/builtin/romfs/lang/de_DE.json:
```diff
@@ -668,6 +668,13 @@
         "hex.builtin.view.hex_editor.search.find": "Suchen",
         "hex.builtin.view.hex_editor.search.hex": "Hex",
         "hex.builtin.view.hex_editor.search.string": "String",
+        "hex.builtin.view.hex_editor.search.string.encoding": "Kodierung",
```
Only add translations for the languages you actually know. It's better to have no translation (and just fallback to english) than having extremely cryptic translations

WerWolv commented on plugins/builtin/include/content/popups/hex_editor/popup_hex_editor_find.hpp:
```diff
@@ -0,0 +1,52 @@
+#include <optional>
+#include <string>
+
+using namespace hex;
```
Never use `using namespace` in a header or global scope

WerWolv commented on plugins/builtin/source/content/popups/hex_editor/popup_hex_editor_find.cpp:
```diff
@@ -0,0 +1,227 @@
+#include <hex/helpers/crypto.hpp>
+#include <codecvt>
+
+using namespace hex;
```
Everything here should be encapsulated in the `hex::plugin::builtin` namespace instead of doing this

WerWolv commented on plugins/builtin/source/content/popups/hex_editor/popup_hex_editor_find.cpp:
```diff
@@ -0,0 +1,227 @@
+            }
+            this->drawButtons();
+            constexpr static std::array StringModeEncodingNames = {
+                "UTF-8",
```
Use the localization strings you've added here

WerWolv commented on plugins/builtin/source/content/popups/hex_editor/popup_hex_editor_find.cpp:
```diff
@@ -0,0 +1,227 @@
+                    ImGui::EndCombo();
+            }
+            constexpr static std::array StringModeEndianessNames = {
+                    "Little endian",
```
Same here. I think there's even some translations in the ui plugin library that you can use. `hex.ui.common.something`

WerWolv commented on plugins/builtin/source/content/popups/hex_editor/popup_hex_editor_find.cpp:
```diff
@@ -0,0 +1,227 @@
+    if (ImGui::BeginTabBar("##find_tabs")) {
+        if (ImGui::BeginTabItem("hex.builtin.view.hex_editor.search.hex"_lang)) {
+            m_isStringMode = false;
+            if (ImGuiExt::InputTextIcon("##input", ICON_VS_SYMBOL_NUMERIC, m_input, ImGuiInputTextFlags_EnterReturnsTrue | ImGuiInputTextFlags_AutoSelectAll | ImGuiInputTextFlags_CharsHexadecimal)) {
```
I usually like to add new lines before/after control flow statements and "logical block" to make things a bit more readable.

WerWolv commented on plugins/builtin/source/content/popups/hex_editor/popup_hex_editor_find.cpp:
```diff
@@ -0,0 +1,227 @@
+                    else if (m_stringModeEndianess == StringModeEndianness::Little)
+                        nativeEndianess = std::endian::little;
+                        
+                    auto swapEndianess = [](auto &value, std::endian endian, StringModeEncoding encoding) {
```
There's byteswap utilities in `<hex/helpers/utils.hpp>`. No need to use intrinsics

PerikiyoXD commented:
To clarify changes, I've refactored the code a bit, extracted the popup code to it's own header and source file, extracted the input processing to only trigger when needed and tweaked the UI a bit.

WerWolv commented on plugins/builtin/source/content/popups/hex_editor/popup_hex_editor_find.cpp:
```diff
@@ -0,0 +1,296 @@
+            //const static auto SearchPositionFormat = "hex.builtin.view.hex_editor.search.position"_lang;
+            // // Status label // std::atomic<bool> m_reachedEnd   = false;
+            if (m_reachedEnd) {
+                auto statusString = "No more results";
```
Missing translation here

WerWolv commented on plugins/builtin/source/content/popups/hex_editor/popup_hex_editor_find.cpp:
```diff
@@ -0,0 +1,296 @@
+                m_searchBackwards = false;
+            }
+
+            // TODO: Implement this
```
This TODO looks outdated? The functionality seems to be implemented but there's some commented code here that should be removed

WerWolv commented on plugins/builtin/source/content/popups/hex_editor/popup_hex_editor_find.cpp:
```diff
@@ -0,0 +1,296 @@
+    }
+
+    void PopupFind::drawTabContents() {
+        static std::array EncodingNames = {
```
These two arrays can't be `static`. Otherwise the translation will only be initialized the first time and then never be updated again, even if the language changes

WerWolv commented on plugins/builtin/source/content/popups/hex_editor/popup_hex_editor_find.cpp:
```diff
@@ -0,0 +1,296 @@
+            if (ImGui::BeginCombo("hex.builtin.view.hex_editor.search.string.encoding"_lang,
+                                  EncodingNames[std::underlying_type_t<Encoding>(m_stringEncoding.load())])) {
+                if (ImGui::Selectable(EncodingNames[0],
+                                      m_stringEncoding == Encoding::UTF8)) {
```
Personally I'd remove that new line, it makes it a bit hard to read

WerWolv commented on plugins/builtin/source/content/popups/hex_editor/popup_hex_editor_find.cpp:
```diff
@@ -0,0 +1,296 @@
+            ImGui::BeginDisabled(m_stringEncoding == Encoding::UTF8);
+            {
+                if (ImGui::BeginCombo("hex.builtin.view.hex_editor.search.string.endianness"_lang,
+                                      EndiannessNames[std::underlying_type_t<Endianness>(m_stringEndianness.load())])) {
```
`std::to_underlying(...)`

WerWolv commented on plugins/builtin/source/content/popups/hex_editor/popup_hex_editor_find.cpp:
```diff
@@ -0,0 +1,296 @@
+                break;
+        }
+    }
+} // namespace hex::plugin::builtin
```
While this is a decent comment, this would be the only one in the entire project
